internal/host/plugin/repository_host_set.go

```diff
@@ -152,6 +156,364 @@ func (r *Repository) CreateSet(ctx context.Context, scopeId string, s *HostSet,
+					PreferredEndpoint: &store.PreferredEndpoint{
+						HostSetId: newSet.PublicId,
+						Priority:  1,
+					},
+				}
+				_, err := w.Delete(ctx, pep, db.WithWhere("host_set_id = ?", newSet.PublicId), db.NewOplogMsg(&peDeleteOplogMsg))
+				if err != nil {
+					return errors.Wrap(ctx, err, op)
+				}
+
+				// Only append the oplog message if an operation was actually
+				// performed.
+				if peDeleteOplogMsg.Message != nil {
+					if !recordUpdated {
+						// we only updated secrets, so we need to increment the
+						// version of the host catalog manually.
+						returnedSet = newSet.clone()
+						returnedSet.Version = uint32(version) + 1
+						var hsOplogMsg oplog.Message
+						numUpdated, err := w.Update(
+							ctx,
+							returnedSet,
+							[]string{"version"},
+							[]string{},
+							db.NewOplogMsg(&hsOplogMsg),
+							db.WithVersion(&version),
+						)
+						if err != nil {
+							return errors.Wrap(ctx, err, op)
+						}
+
+						if numUpdated != 1 {
+							return errors.New(ctx, errors.MultipleRecords, op, fmt.Sprintf("expected 1 set to be updated, got %d", numUpdated))
+						}
+
+						recordUpdated = true
+						msgs = append(msgs, &hsOplogMsg)
+					}
+
+					msgs = append(msgs, &peDeleteOplogMsg)
+				}
+
+			case endpointOpUpdate:
+				// Delete all old endpoint entries.
+				var peDeleteOplogMsg oplog.Message
+				pep := &host.PreferredEndpoint{
+					PreferredEndpoint: &store.PreferredEndpoint{
+						HostSetId: newSet.PublicId,
+						Priority:  1,
+					},
+				}
+				_, err := w.Delete(ctx, pep, db.WithWhere("host_set_id = ?", newSet.PublicId), db.NewOplogMsg(&peDeleteOplogMsg))
+				if err != nil {
+					return errors.Wrap(ctx, err, op)
+				}
+
+				// Only append the oplog message if an operation was actually
+				// performed.
+				if peDeleteOplogMsg.Message != nil {
+					msgs = append(msgs, &peDeleteOplogMsg)
+				}
+
+				// Create the new entries.
+				peCreateOplogMsgs := make([]*oplog.Message, 0, len(preferredEndpoints))
+				if err := w.CreateItems(ctx, preferredEndpoints, db.NewOplogMsgs(&peCreateOplogMsgs)); err != nil {
+					return err
+				}
+
+				if !recordUpdated {
+					// we only updated secrets, so we need to increment the
+					// version of the host catalog manually.
+					returnedSet = newSet.clone()
+					returnedSet.Version = uint32(version) + 1
+					var hsOplogMsg oplog.Message
+					numUpdated, err := w.Update(
+						ctx,
+						returnedSet,
+						[]string{"version"},
+						[]string{},
+						db.NewOplogMsg(&hsOplogMsg),
+						db.WithVersion(&version),
+					)
+					if err != nil {
+						return errors.Wrap(ctx, err, op)
+					}
+
+					if numUpdated != 1 {
+						return errors.New(ctx, errors.MultipleRecords, op, fmt.Sprintf("expected 1 set to be updated, got %d", numUpdated))
+					}
+
+					recordUpdated = true
+					msgs = append(msgs, &hsOplogMsg)
+				}
+
+				msgs = append(msgs, peCreateOplogMsgs...)
+			}
+
+			if !pluginCalledSuccessfully {
```

Comment:
Move this to be the last thing that happens in the transaction.

Reply:
Done!